### ISSUE TITLE
`Vec` reallocation failed (our`memcpy` bug?)

### ISSUE THREAD

**Issue by greenhat** [0xMiden/compiler#811]
Discovered in https://github.com/0xMiden/compiler/pull/808

To reproduce add the following test to `ests/integration/src/rust_masm_tests/misc.rs`:
```rust
/// Test that Vec reallocation correctly copies existing elements to the new buffer.
///
/// This test reproduces a bug in Vec reallocation. The test pushes 5 elements to a Vec (which
/// triggers reallocation from capacity 4 to 8), then verifies all elements are accessible.
#[test]
fn test_vec_realloc_copies_data() {
    let main_fn = r#"() -> Felt {
        extern crate alloc;
        use alloc::vec::Vec;

        // Create a Vec and push 5 Felt elements
        // Vec typically starts with capacity 4, so the 5th push triggers reallocation
        let mut v: Vec<Felt> = Vec::new();
        v.push(felt!(11111));
        v.push(felt!(22222));
        v.push(felt!(33333));
        v.push(felt!(44444));
        v.push(felt!(55555));  // This triggers reallocation

        // Sum all elements - if realloc doesn't copy, first 4 elements will be garbage
        let sum = v[0] + v[1] + v[2] + v[3] + v[4];

        // Expected: 11111 + 22222 + 33333 + 44444 + 55555 = 166665
        sum
    }"#;

    let config = WasmTranslationConfig::default();
    let mut test = CompilerTestBuilder::rust_fn_body_with_stdlib_sys(
        "vec_realloc_copies_data",
        main_fn,
        config,
        [],
    )
    .build();

    test.expect_wasm(expect_file!["../../expected/vec_realloc_copies_data.wat"]);
    test.expect_ir(expect_file!["../../expected/vec_realloc_copies_data.hir"]);
    test.expect_masm(expect_file!["../../expected/vec_realloc_copies_data.masm"]);

    let package = test.compiled_package();

    let args: [Felt; 0] = [];
    eval_package::<Felt, _, _>(&package, [], &args, &test.session, |trace| {
        let result: u64 = trace.parse_result::<Felt>().unwrap().as_int();
        // Expected sum: 11111 + 22222 + 33333 + 44444 + 55555 = 166665
        // Result: The test fails with result 55658 instead of 166665, confirming that:
        // The 5th element (55555) is stored correctly (55658 - 55555 = 103 came from some values
        // at the first four elements)
        assert_eq!(result, 166665, "Vec reallocation failed to copy existing elements");
        Ok(())
    })
    .unwrap();
}
```

Our `memcpy` is at https://github.com/0xMiden/compiler/blob/ba60fd685477b9893e6af073cd47dcd85b8060f0/codegen/masm/src/emit/mem.rs?plain=1#L670-L816
